File: swift/Sources/Autograph/KeyPair.swift
import Clibautograph
import Foundation

private func createKeyPair() -> Bytes {
    createBytes(autograph_key_pair_size())
}

public func generateIdentityKeyPair() throws -> Bytes {
    var keyPair = createKeyPair()
    let success = autograph_identity_key_pair(&keyPair)
    if !success {
        throw Error.keyPair
    }
    return keyPair
}

public func generateKeyPair() throws -> Bytes {
    var keyPair = createKeyPair()
    let success = autograph_key_pair(&keyPair)
    if !success {
        throw Error.keyPair
    }
    return keyPair
}

File: android/src/main/java/sh/autograph/KeyPair.kt
package sh.autograph

class KeyPair {
    companion object {
        init {
            System.loadLibrary("autograph")
        }

        private external fun autographKeyPairSize(): Int

        private external fun autographEphemeralKeyPair(keyPair: ByteArray): Boolean

        private external fun autographIdentityKeyPair(keyPair: ByteArray): Boolean

        private fun createKeyPair(): ByteArray = ByteArray(autographKeyPairSize())

        fun generateKeyPair(): ByteArray {
            val keyPair = createKeyPair()
            val success = autographEphemeralKeyPair(keyPair)
            if (!success) {
                throw RuntimeException("Key generation failed")
            }
            return keyPair
        }

        fun generateIdentityKeyPair(): ByteArray {
            val keyPair = createKeyPair()
            val success = autographIdentityKeyPair(keyPair)
            if (!success) {
                throw RuntimeException("Key generation failed")
            }
            return keyPair
        }
    }
}
